src/merkle_tree/mod.rs

```diff
@@ -181,6 +183,212 @@ impl<P: Config> Path<P> {
+    pub leaf_indexes: Vec<usize>,
+}
+
+impl<P: Config> MultiPath<P> {
+    /// Returns a compressed MultiPath containing multiple encoded authentication paths for `indexes`
+    fn compress(
+        indexes: impl IntoIterator<Item = usize>,
+        leaf_siblings_hashes: Vec<P::LeafDigest>,
+        auth_paths: Vec<Path<P>>,
+    ) -> Result<Self, crate::Error> {
+        let indexes = Vec::from_iter(indexes);
+        // use multipath for more than 1 leaf
+        assert!(
+            indexes.len() > 1,
+            "Expected more than one leaf to verify for MultiPath, got {}",
+            indexes.len()
+        );
+
+        let mut auth_paths_prefix_lenghts: Vec<usize> = Vec::with_capacity(indexes.len());
+        let mut auth_paths_suffixes: Vec<Vec<P::InnerDigest>> = Vec::with_capacity(indexes.len());
+
+        auth_paths_prefix_lenghts.push(0);
+        auth_paths_suffixes.push(auth_paths[0].auth_path.clone());
+
+        let mut prev_path = auth_paths[0].clone();
+
+        // Encode paths with Incremental Encoding (Front compression)
+        for path in auth_paths[1..].to_vec() {
+            let mut prefix_len = 0;
+            if prev_path.auth_path.len() != 0 && path.auth_path.len() != 0 {
+                while prev_path.auth_path[prefix_len] == path.auth_path[prefix_len] {
+                    prefix_len += 1;
+                    if prefix_len == prev_path.auth_path.len() || prefix_len == path.auth_path.len()
+                    {
+                        break;
+                    }
+                }
+            }
+            auth_paths_prefix_lenghts.push(prefix_len);
+            if prefix_len == prev_path.auth_path.len() {
+                auth_paths_suffixes.push(vec![]);
+            } else {
+                auth_paths_suffixes.push(path.auth_path[prefix_len..].to_vec().clone());
+            }
+            prev_path = path;
+        }
+
+        Ok(MultiPath {
+            leaf_indexes: indexes,
+            auth_paths_prefix_lenghts,
+            auth_paths_suffixes,
+            leaf_siblings_hashes,
+        })
+    }
+
+    /// Returns the decompressed authentication paths for every leaf index
+    fn decompress(
+        &'_ self,
+    ) -> Result<impl '_ + Iterator<Item = Vec<P::InnerDigest>>, crate::Error> {
+        // Incrementally reconstruct all the paths
+        let mut curr_path = self.auth_paths_suffixes[0].clone();
+        let mut auth_paths = (0..self.leaf_indexes.len())
+            .map(|_| Vec::with_capacity(curr_path.len()))
+            .collect::<Vec<Vec<P::InnerDigest>>>();
+        auth_paths[0] = self.auth_paths_suffixes[0].clone();
+
+        for i in 1..auth_paths.len() {
+            auth_paths[i].extend_from_slice(&curr_path[0..self.auth_paths_prefix_lenghts[i]]);
+            auth_paths[i].extend(self.auth_paths_suffixes[i].clone());
+            curr_path = auth_paths[i].clone();
+        }
+        Ok(auth_paths.into_iter())
+    }
```

Comment:
These are effectively two extra loops that are not needed (i think).
Good that you have benches, because now you can test whether the following will give you any speedup:

Combining these into a single loop that avoids pre-mature allocation rather than creating empty vec of vectors and mutating each in place again in the next loop, why not start with an empty `let mut auth_paths = Vec::with_capacity(something)` and just push newly created `auth_path`?
Then you can avoid `extend_from_slice` followed directly by an `extend`